android/app/src/main/java/app/covidshield/module/ExposureNotificationModule.kt

```diff
@@ -1,63 +1,103 @@
 import app.covidshield.extensions.toWritableArray
 import app.covidshield.extensions.toWritableMap
+import app.covidshield.models.Configuration
 import com.facebook.react.bridge.Promise
 import com.facebook.react.bridge.ReactApplicationContext
 import com.facebook.react.bridge.ReactContextBaseJavaModule
 import com.facebook.react.bridge.ReactMethod
 import com.facebook.react.bridge.ReadableArray
 import com.facebook.react.bridge.ReadableMap
 import com.google.android.gms.nearby.Nearby
+import java.io.File
+import java.util.UUID
+
+private const val SUMMARY_HIDDEN_KEY = "_summaryIdx"
 
 class ExposureNotificationModule(context: ReactApplicationContext) : ReactContextBaseJavaModule(context) {
 
-    private val exposureNotificationClient = Nearby.getExposureNotificationClient(context.applicationContext)
+    private val exposureNotificationClient by lazy {
+        Nearby.getExposureNotificationClient(context.applicationContext)
+    }
 
     override fun getName(): String = "ExposureNotification"
 
     @ReactMethod
     fun start(promise: Promise) {
-        promise.resolve(null)
+        exposureNotificationClient.start().bindPromise(promise) {
+            resolve(Unit)
+        }
     }
 
     @ReactMethod
     fun stop(promise: Promise) {
-        exposureNotificationClient.stop()
-            .addOnCompleteListener {
-                promise.resolve(null)
-            }
+        exposureNotificationClient.stop().bindPromise(promise) {
+            resolve(Unit)
+        }
     }
 
     @ReactMethod
     fun resetAllData(promise: Promise) {
+        // This method does not exist in the android nearby SDK.
         promise.resolve(null)
     }
 
     @ReactMethod
     fun getStatus(promise: Promise) {
-        exposureNotificationClient.isEnabled
-            .addOnSuccessListener {
-                val status = if (it) Status.ACTIVE else Status.DISABLED
-                promise.resolve(status.value)
-            }
-            .addOnFailureListener {
-                promise.resolve(Status.DISABLED.value)
-            }
+        exposureNotificationClient.isEnabled.bindPromise(promise, Status.DISABLED.value) { isEnabled ->
+            val status = if (isEnabled) Status.ACTIVE else Status.DISABLED
+            resolve(status.value)
+        }
     }
 
     @ReactMethod
     fun detectExposure(configuration: ReadableMap, diagnosisKeysURLs: ReadableArray, promise: Promise) {
-        promise.resolve(emptyMap<String, String>().toWritableMap())
+        promise.rejectOnException {
+            val exposureConfiguration = configuration.parse(Configuration::class.java).toExposureConfiguration()
+            val files = diagnosisKeysURLs.parse(String::class.java).map { File(it) }
+            val token = UUID.randomUUID().toString()
+            exposureNotificationClient
+                .provideDiagnosisKeys(files, exposureConfiguration, token)
+                .continueWithTask { exposureNotificationClient.getExposureSummary(token) }
```

Comment:
@henrytao-me I'm not familiar with the Task API, does this chain failures like promises too? Otherwise we might have a problem.

Reply:
Yes. It chains like promise. If the first chain is failed, it will skip the next one and go to the end.